src/compiler/checker.ts

```diff
@@ -573,7 +573,16 @@ namespace ts {
                         declarationNameToString(propertyName), typeof nameArg === "string" ? nameArg : declarationNameToString(nameArg));
                     return undefined;
                 }
-                if (result.flags & SymbolFlags.BlockScopedVariable) {
+
+                // Only check for block-scoped variable if we are looking for the name with variable meaning
+                // For example,
+                //      declare module foo {
+                //          interface bar {}
+                //      }
+                // let foo/*1*/: foo/*2*/.bar;
+                // The foo at /*1*/ and /*2*/ will share same symbol with two meaning block-scope variable and namespace module.
+                // However, only when we try to resolve name in /*1*/ which is used in variable position, we want to check for block- scoped
+                if (meaning & SymbolFlags.Variable && result.flags & SymbolFlags.BlockScopedVariable) {

```

Comment:
I would actually check `meaning & SymbolFlags.BlockScopeVariable` instead of Variable. I think it is more direct. It's basically saying "if I will accept a result that is at least a block scoped variable, and the thing I found is at least a block scoped variable, then I found this symbol *because* it was a block scoped variable". The condition you have right now would be true (and should be false) in the hypothetical case where you were only looking for function scoped variables.